README.md
```
Copyright (C) 2013, 2014, 2015, 2016 Digital Freedom Foundation
  This file is part of GNUKhata:A modular,robust and Free Accounting System.

  GNUKhata is Free Software; you can redistribute it and/or modify
  it under the terms of the GNU Affero General Public License as
  published by the Free Software Foundation; either version 3 of
  the License, or (at your option) any later version.

  GNUKhata is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU Affero General Public License for more details.

  You should have received a copy of the GNU Affero General Public
  License along with GNUKhata (COPYING); if not, write to the
  Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
  Boston, MA  02110-1301  USA59 Temple Place, Suite 330,
```
Contributors:
"Krishnakant Mane" <kk@gmail.com>
### Installation
<!--
To install GNUKhata in development mode, there are a few things to understand.
First and for most, GNUKhata core is needed for which you will be building the web app.
Secondly, All dependencies related to Python based code will be installed in a virtual environment.
A virtual environment isolates your own personal copy of the Python programming language,
Also gives you freedom to have several copies and all this never affects the main Python setup.
In addition keep in mind that the core logic needs Postgresql database and a few other things to work.
As a web app developer, you may not need to know much about the core but knowing how to run it will be necessary.
We will first install the dependencies, then create the virtual environment and add the necessary Python libraries, incuding the Pyramid framework.
Then we clone our copy of gkcore and gkwebapp.
Note that to install libraries and to run GNUKhata, the virtual environment must be first activated.
Note also that steps given here should be followed completely.  If any step is skipped it might result in GNUKhata not working at all or working improperly.
BE CAREFULL, You ARe WARNED!
All the following set of commands should be given in a terminal.
You can press alt + ctrl + t to get a terminal.
you will see the $ prompt as soon as the terminal opens.
-->

### Note: you have to press enter after completely typing every command.
### Note: the `$` sign is just provided to indicate a terminal, so please don't include it in your commands.

* On debian/Ubuntu distributions Install python-virtualenv postgresql and dependencies using following command 

> `$ sudo apt-get install python3-virtualenv postgresql python3-dev libpq-dev git python3-setuptools build-essential` 


* Create a python virtualenv in a directory(NOT IN gkwebapp or gkcore. If using emacs please do so in '.virtualenvs' directory in home.) using:

> `$ virtualenv gkenv `

* change directory to gkenv:

> `$ cd gkenv`

* Activate your virtualenv using:

> `source bin/activate`

* Fork gkcore from https://gitlab.com/gnukhata/gkcore

* Create and add your SSH key to gitlab by following this guide - https://gitlab.com/help/gitlab-basics/create-your-ssh-keys.md

* Clone gkcore in your workspace using:

> `git clone git@gitlab.com:<username>/gkcore.git`

* Change permission of gkcore to gain write acess.

> `sudo chmod 775 gkcore`

* Change to gkcore directory. Look inside the directory. You must find files like `gkutil.sh`, `setup.py`, `initdb.py`

* Give permission to gkutil.sh file to execute and execute the same using:

> `chmod 755 gkutil.sh`

> `./gkutil.sh`

* Activate your virtual environment and run setup.py using:

> `python3 setup.py develop`

* Your environment will be checked for all the required libraries and the missing ones will be downloaded.

* Now we have to run initdb.py script. This will create tables in our database. To run this script we need to switch to a user 'gkadmin' which was created when we ran 'gkutil.sh' script.

> `sudo su gkadmin`

Activate your virtualenv and then run initdb.py.

> `python initdb.py`

* To run gkcore server in development mode use:

> `pserve development.ini`

gkcore is now accessible at `http://localhost:6543`🎉 

<!--
* Open another terminal and change directory to gkwebapp
* Activate virtualenv and run setup.py using

> `python setup.py develop`

* To run gkwebapp server in development mode use:

> `pserve development.ini`
Documentation
=============
gkcore is the core engine for GNUKhata <gnukhata.in> a free and open source accounting/ book keeping software.
The core engine contains the database creation and management code along with the code for implementing the logic in form of RESTful API.
To get the code running on your machine as developers, you need to create a virtual environment of Python and then create the databaes and it's dedicated users.

NOTE: PLEASE ENTER ALL COMMANDS AS THEY HAVE BEEN GIVEN INCLUDING QUOTES ("")
These are the steps to get the database initialised.
WARNING: "perform these commands with the full knowledge of what you are doing "
1, firstly we need a system user so issue the command sudo useradd gkadmin and press enter
2, create a role with same name: type sudo -u postgres psql -c "create role gkadmin with login"
3, grant all privileges for this do:
a: sudo -u postgres psql -c "alter role gkadmin createdb;"
b: sudo -u postgres psql -c "grant all privileges on database template1 to gkadmin;"
4, create the database, issue command sudo -u postgres psql -c "create database gkdata"
-->


CHANGELOG
[16 December 2019 Release 6.50]
* Function to cancel delivery note.
* New table to store cancelled delivery note data.
* Function to cancel invoice when delivery note is associated with the invoice.
* New fields in product table  to save discount in percent and as amount. 
* New field discflag in table invoice , invoicebin and delchal.
* Field to save narration in purchase and sales order.
* Change in APIs related to the field added. 
* Enhancement in code to calculate discount in Invoice, Cashmemo and Delivery note.
* Function to show states of customers for which gstin is provided.
* Performance Enhancements in APIs and database.
* Bug Fixes.



[16 August 2019 Release 6.0 Updates] 
* Round Off feature for Delivery note , Credit/Debit note and Purchase/Sales order with automatic accounting entries.
* Accounts necessary for automatic accounting entries are created atomatically if they are found to be absent.
* PIN Code of Customer/Supplier is saved.
* AccountsByRule function tweaked to allow both customers and suppliers in Sale and Purchase vouchers.
* P&L and Balance Sheet reports API accepts any range of date within a financial year.
* Provision to have multiple sales invoice number series.
* Field for eway bill has been added.
* Field for narration in Invoice and Debit Credit Note.
* Default accounts bug resolved.
* Dashboard performance enhancement.
* Performance Enhancements in APIs and database.
* Bug Fixes.


[26 April 2019 Release 6.0]
* Added API for dashboard and made related modifications in other APIs.
* Added API for Budget.
* Modified DELETE API of Invoice to implement Cancellation of Invoice.
* Automatic voucher number enhanced.
* Modifications in API of Invoice for Rounding Off invoice total value.
* Accounting with Invoice option depricated.
* Organisation POST API returns orgcode.
* Modification in tax API to allow Godown In Charge to add tax to product.
* Performance Enhancements.

	
[29-November-2018 Release 5.50]
* API for GSTR1 b2b, b2cl, b2cs, cdnr, cdnur, hsn1.  
* Modifications in Profit and Loss Report.  
* New fields for Sale Price and MRP in Product Table.  
* Corrected query that returns invid after Invoice/Cash Memo is saved.  
* Function to delete invoices that have nothing linked to them and no billwise entry made.  
* Function to get list of invoice that can be rectified.
* Improved documentation.	
	
[15-August-2018 Release 5.25]
* New Unit of measurement.
* Bug fixes and enhancements for bill wise accounting.
* Connection string for linux and windows.	

[15-June-2018 Release 5.10]
* Api for automatic sales and purchase voucher entry	
* Api for automatic credit & debit note voucher entry
* Payment and receipt to be made by bank and cash transfer
* Sorting functionality in major reports	
* Automatic gst sammary report
* gkupgrade() revamp
* new flags in organisation
* Default accounts for sales, purchase,bank,cash transaction
* Calculation enhancement in Profit and Loss report and Balance sheet


[12-April-2018 Release 5.0]
* Roll over without close book	
* Debit credit note API
* Rejection Note API	
* GST compliant Sales and purchase order API
* Subgroups under Direct Income and Direct Expenses
* Altered rules in accounts by rule API
* Added new columns "invoicetotal","address" in invoice table
* Profit and loss statement code changed for structured tree format
* Free quantity and billed quantity also added when stock details   	
* List of outstanding invoices.
* GST compliant sales and purchase register
* Inoutflag for cashmemo and delivery chalan
* New columns added in delivery note to make it GST compliant
* GST summary report

	
[10-January-2018 Release 4.50]
* Refinement in Transaction API
* Function in Transaction API to auto generate Voucher No.
* Function in Bill Wise API that returns all unpaid invoices.
* Function in Bill Wise API that returns both unpaid and partially paid invoices.
* Function in User API that returns list of users having same role and their role.
* Function in User API that returns all data of a user(accessible only to admin).
* All details of a user can be edited by admin.
* Consignee details can be added in Delivery Note.
* Function in Godowns API to return name and address of all godowns that contain products.
* Modified PUT of Invoice API.

[10-November-2017 Release 4.25 Update]
* Permissions to Godown In Charge modified.
* CESS field added in invoice and cash memo.
* Sales and purchase register now reflects along with CESS.
* Gkupgrade modified as per modifications in gkdb.
* Transfer Note get all function modified.
* A function to get user role.
* Get all delchal function.
* Minor bugs solved.

[21-September-2017 Release 4.25]
*Gkupgrade modified as per modifications in gkdb
*List of users report
*Function in api_godown modified to get number of product a content.
*Api for billwise accounting
*List of vouchers in chronological order.
*Get function in api_log modified to get godowns associated with a godown in charge and sr no.
*Edit function of category modified to edit/delete/added child categories tax and specifications when its parent category's parameters are edited.
*Godown in charge user access corrected.
*Function to get customer details when account code is recieved and matches the same.
*Fromgodown column added in transfernote table and api also revamped.
*Function corrected to get godown wise product.
*Altered sales and purchase rules so that new sub-groups can also be added.
*Category wise stock on hand report.
*Sales and purchase registers.
*Modified unbilled deliveries function.
*Logo column added in organisation also api modified accordingly.
*Goprod table and functions related to it modified.
*Bank name,branch name,instrumentno,instrumentdate to store Bank details [Instrument details] in vouchers also functions get and post changed accordingly.
*List of transfernotes report.
*New columns in organisation table invsflag and billwiseflag to set Organisation preferences.
*List of invoices report.

initdb.py

"""
Copyright (C) 2013, 2014, 2015, 2016 Digital Freedom Foundation
Copyright (C) 2017, 2018 Digital Freedom Foundation & Accion Labs 
  This file is part of GNUKhata:A modular,robust and Free Accounting System.

  GNUKhata is Free Software; you can redistribute it and/or modify
  it under the terms of the GNU Affero General Public License as
  published by the Free Software Foundation; either version 3 of
  the License, or (at your option) any later version.and old.stockflag = 's'

  GNUKhata is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU Affero General Public License for more details.

  You should have received a copy of the GNU Affero General Public
  License along with GNUKhata (COPYING); if not, write to the
  Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
  Boston, MA  02110-1301  USA59 Temple Place, Suite 330,


Contributors:
"Krishnakant Mane" <kk@gmail.com>
"Ishan Masdekar " <imasdekar@dff.org.in>
"Navin Karkera" <navin@dff.org.in>
"""

from gkcore.models.meta import dbconnect
from gkcore.models.gkdb import metadata
from gkcore.models import gkdb
from sqlalchemy.sql import select
from sqlalchemy import func
import datetime
from time import strftime

"""
This module is used only once per installation.
It will use the sqlalchemy's create_all function to convert all python based table spects to real sql tables.
Refer to gkdb.py in models package for structure of all tables expressed in the alchemy expression language.
After creating all tables, it will also create the signature based on timestamp and store in the database.
"""
eng = dbconnect()
metadata.create_all(eng)
print "database created successfully"
eng.execute("alter table groupsubgroups add  foreign key (subgroupof) references groupsubgroups(groupcode)")
eng.execute("alter table categorysubcategories add  foreign key (subcategoryof) references categorysubcategories(categorycode)")
eng.execute("alter table unitofmeasurement add  foreign key (subunitof) references unitofmeasurement(uomid)")

try:
    uomscount = eng.execute(select([func.count(gkdb.unitofmeasurement.c.uomid).label("numofuom")]))
    numofuom = uomscount.fetchone()
    if int(numofuom["numofuom"]) == 0:
        dictofuqc = {'BAG':'BAG','BGS':'BAGS','BLS':'BAILS','BTL':'BOTTLES','BOU':'BOU','BOX':'BOXES','BKL':'BUCKLES','BLK':'BULK','BUN':'BUNCHES','BDL':'BUNDLES','CAN':'CANS','CTN':'CARTONS','CAS':'CASES','CMS':'CENTIMETER','CHI':'CHEST','CLS':'COILS','COL':'COLLIES','CRI':'CRATES','CCM':'CUBIC CENTIMETER','CIN':'CUBIC INCHES','CBM':'CUBIC METER','CQM':'CUBIC METERS','CYL':'CYLINDER','SDM':'DECAMETER SQUARE','DAY':'DAYS','DOZ':'DOZEN','DRM':'DRUMS','FTS':'FEET','FLK':'FLASKS','GMS':'GRAMS','TON':'GREAT BRITAIN TON','GGR':'GREAT GROSS','GRS':'GROSS','GYD':'GROSS YARDS','HBK':'HABBUCK','HKS':'HANKS','HRS':'HOURS','INC':'INCHES','JTA':'JOTTA','KGS':'KILOGRAMS','KLR':'KILOLITER','KME':'KILOMETERS','LTR':'LITERS','LOG':'LOGS','LOT':'LOTS','MTR':'METER','MTS':'METRIC TON','MGS':'MILLIGRAMS','MLT':'MILLILITER','MMT':'MILLIMETER','NONE':'NOT CHOSEN','NOS':'NUMBERS','ODD':'ODDS','PAC':'PACKS','PAI':'PAILS','PRS':'PAIRS','PLT':'PALLETS','PCS':'PIECES','LBS':'POUNDS','QTL':'QUINTAL','REL':'REELS','ROL':'ROLLS','SET':'SETS','SHT':'SHEETS','SLB':'SLABS','SQF':'SQUARE FEET','SQI':'SQUARE INCHES','SQC':'SQUARE CENTIMETERS','SQM':'SQUARE METER','SQY':'SQUARE YARDS','BLO':'STEEL BLOCKS','TBL':'TABLES','TBS':'TABLETS','TGM':'TEN GROSS','THD':'THOUSANDS','TIN':'TINS','TOL':'TOLA','TRK':'TRUNK','TUB':'TUBES','UNT':'UNITS','UGS':'US GALLONS','VLS':'VIALS','CSK':'WOODEN CASES','YDS':'YARDS'}

        for unit, desc in dictofuqc.items():
            eng.execute("insert into unitofmeasurement(unitname, conversionrate, description, sysunit)values('%s',0.00,'%s',1)"%(unit,desc))
            dictofuqc.pop(unit,0)

    statescount = eng.execute(select([func.count(gkdb.state.c.statecode).label("numberofstates")]))
    numberofstates = statescount.fetchone()
    if int(numberofstates["numberofstates"]) == 0:
        eng.execute("insert into state( statecode, statename, abbreviation)values(1, 'Jammu and Kashmir', 'JK')")
        eng.execute("insert into state( statecode, statename, abbreviation)values(2, 'Himachal Pradesh', 'HP')")
        eng.execute("insert into state( statecode, statename, abbreviation)values(3, 'Punjab', 'PB')")
        eng.execute("insert into state( statecode, statename, abbreviation)values(4, 'Chandigarh', 'CH')")
        eng.execute("insert into state( statecode, statename, abbreviation)values(5, 'Uttarakhand', 'UK')")
        eng.execute("insert into state( statecode, statename, abbreviation)values(6, 'Haryana', 'HR')")
        eng.execute("insert into state( statecode, statename, abbreviation)values(7, 'Delhi', 'DL')")
        eng.execute("insert into state( statecode, statename, abbreviation)values(8, 'Rajasthan', 'RJ')")
        eng.execute("insert into state( statecode, statename, abbreviation)values(9, 'Uttar Pradesh', 'UP')")
        eng.execute("insert into state( statecode, statename, abbreviation)values(10, 'Bihar', 'BR')")
        eng.execute("insert into state( statecode, statename, abbreviation)values(11, 'Sikkim', 'SK')")
        eng.execute("insert into state( statecode, statename, abbreviation)values(12, 'Arunachal Pradesh', 'AR')")
        eng.execute("insert into state( statecode, statename, abbreviation)values(13, 'Nagaland', 'NL')")
        eng.execute("insert into state( statecode, statename, abbreviation)values(14, 'Manipur', 'MN')")
        eng.execute("insert into state( statecode, statename, abbreviation)values(15, 'Mizoram', 'MZ')")
        eng.execute("insert into state( statecode, statename, abbreviation)values(16, 'Tripura', 'TR')")
        eng.execute("insert into state( statecode, statename, abbreviation)values(17, 'Meghalaya', 'ML')")
        eng.execute("insert into state( statecode, statename, abbreviation)values(18, 'Assam', 'AS')")
        eng.execute("insert into state( statecode, statename, abbreviation)values(19, 'West Bengal', 'WB')")
        eng.execute("insert into state( statecode, statename, abbreviation)values(20, 'Jharkhand', 'JH')")
        eng.execute("insert into state( statecode, statename, abbreviation)values(21, 'Odisha', 'OR')")
        eng.execute("insert into state( statecode, statename, abbreviation)values(22, 'Chhattisgarh', 'CG')")
        eng.execute("insert into state( statecode, statename, abbreviation)values(23, 'Madhya Pradesh', 'MP')")
        eng.execute("insert into state( statecode, statename, abbreviation)values(24, 'Gujarat', 'GJ')")
        eng.execute("insert into state( statecode, statename, abbreviation)values(25, 'Daman and Diu', 'DD')")
        eng.execute("insert into state( statecode, statename, abbreviation)values(26, 'Dadra and Nagar Haveli', 'DH')")
        eng.execute("insert into state( statecode, statename, abbreviation)values(27, 'Maharashtra', 'MH')")
        eng.execute("insert into state( statecode, statename, abbreviation)values(28, 'Andhra Pradesh', 'AP')")
        eng.execute("insert into state( statecode, statename, abbreviation)values(29, 'Karnataka', 'KA')")
        eng.execute("insert into state( statecode, statename, abbreviation)values(30, 'Goa', 'GA')")
        eng.execute("insert into state( statecode, statename, abbreviation)values(31, 'Lakshdweep', 'LD')")
        eng.execute("insert into state( statecode, statename, abbreviation)values(32, 'Kerala', 'KL')")
        eng.execute("insert into state( statecode, statename, abbreviation)values(33, 'Tamil Nadu', 'TN')")
        eng.execute("insert into state( statecode, statename, abbreviation)values(34, 'Pondicherry', 'PY')")
        eng.execute("insert into state( statecode, statename, abbreviation)values(35, 'Andaman and Nicobar Islands', 'AN')")
        eng.execute("insert into state( statecode, statename, abbreviation)values(36, 'Telangana', 'TS')")
        eng.execute("insert into state( statecode, statename, abbreviation)values(37, 'Andhra Pradesh (New)', 'AP')")
        eng.execute("alter table transfernote add recieveddate date")
    eng.execute("alter table delchal add noofpackages int")
    eng.execute("alter table delchal add modeoftransport text")
except:
    pass



print "secret signature generated"
